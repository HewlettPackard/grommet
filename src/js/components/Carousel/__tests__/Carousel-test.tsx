import React from 'react';
import { render, act, waitFor, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import 'jest-styled-components';
import '@testing-library/jest-dom';

import { Carousel } from '..';

const getSlideOne = () => screen.getByText('Slide One');
const getSlideTwo = () => screen.getByText('Slide Two');
const getSlideThree = () => screen.getByText('Slide Three');

describe('Carousel', () => {
  test('basic', () => {
    const { container } = render(
      <Carousel>
        <div>Slide One</div>
        <div>Slide Two</div>
      </Carousel>,
    );
    expect(getSlideOne()).toBeVisible();
    expect(getSlideTwo()).not.toBeVisible();
    expect(container.firstChild).toMatchSnapshot();
  });

  test('basic with `initialChild: 1`', () => {
    render(
      <Carousel initialChild={1}>
        <div>Slide One</div>
        <div>Slide Two</div>
      </Carousel>,
    );

    expect(getSlideTwo()).toBeVisible();
    expect(getSlideOne()).not.toBeVisible();
  });

  test('arrow navigation: next', async () => {
    render(
      <Carousel controls="arrows">
        <div>Slide One</div>
        <div>Slide Two</div>
        <div>Slide Three</div>
      </Carousel>,
    );

    /**
     * - Currently on "Slide One"
     * - Simulating click on the next arrow button
     * - Expecting "Slide Two" to be visible
     */
    const nextButton = screen.getByLabelText('Next');
    userEvent.click(nextButton);
    await waitFor(() => {
      expect(getSlideOne()).not.toBeVisible();
      expect(getSlideTwo()).toBeVisible();
    });
  });

  test('arrow navigation: previous', async () => {
    render(
      <Carousel controls="arrows">
        <div>Slide One</div>
        <div>Slide Two</div>
        <div>Slide Three</div>
      </Carousel>,
    );

    /**
     * - Currently on "Slide One"
     * - Simulating click on the previous arrow button
     * - Expecting "Slide Three" to be visible
     */
    const previousButton = screen.getByLabelText('Previous');
    userEvent.click(previousButton);
    await waitFor(() => {
      expect(getSlideOne()).not.toBeVisible();
      expect(getSlideThree()).toBeVisible();
    });
  });

  test('selector navigation: forward', async () => {
    render(
      <Carousel controls="selectors">
        <div>Slide One</div>
        <div>Slide Two</div>
        <div>Slide Three</div>
      </Carousel>,
    );

    /**
     * - Currently on "Slide One"
     * - Simulating click on the 3rd selector button
     * - Expecting "Slide Three" to be visible
     */
    const thirdSelector = screen.getAllByRole('button')[2];
    userEvent.click(thirdSelector);
    await waitFor(() => {
      expect(getSlideOne()).not.toBeVisible();
      expect(getSlideThree()).toBeVisible();
    });
  });

  test('selector navigation: backward', async () => {
    render(
      <Carousel initialChild={2} controls="selectors">
        <div>Slide One</div>
        <div>Slide Two</div>
        <div>Slide Three</div>
      </Carousel>,
    );

    /**
     * - Currently on "Slide Three"
     * - Simulating click on the 1st selector button
     * - Expecting "Slide One" to be visible
     */
    const firstSelector = screen.getAllByRole('button')[0];
    userEvent.click(firstSelector);
    await waitFor(() => {
      expect(getSlideThree()).not.toBeVisible();
      expect(getSlideOne()).toBeVisible();
    });
  });

  test('play', async () => {
    jest.useFakeTimers('modern');
    render(
      <Carousel play={500} controls={false}>
        <div>Slide One</div>
        <div>Slide Two</div>
      </Carousel>,
    );

    expect(getSlideOne()).toBeVisible();
    await act(async () => {
      jest.advanceTimersByTime(800);
    });
    await waitFor(() => {
      expect(getSlideOne()).not.toBeVisible();
      expect(getSlideTwo()).toBeVisible();
    });
  });

  test('controlled component', async () => {
    const setActiveSlide = jest.fn();
    const { container } = render(
      <Carousel controls="arrows" activeChild={1} onChild={setActiveSlide}>
        <div>Slide One</div>
        <div>Slide Two</div>
      </Carousel>,
    );

    expect(container.firstChild).toMatchSnapshot();
    expect(setActiveSlide).toHaveBeenCalledWith(1);
  });
});
