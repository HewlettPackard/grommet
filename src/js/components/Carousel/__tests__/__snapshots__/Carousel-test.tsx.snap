// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel basic 1`] = `
<DocumentFragment>
  .c7 {
  display: inline-block;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 24px;
  height: 24px;
  fill: #666666;
  stroke: #666666;
}

.c7 g {
  fill: inherit;
  stroke: inherit;
}

.c7 *:not([stroke])[fill='none'] {
  stroke-width: 0;
}

.c7 *[stroke*='#'],
.c7 *[STROKE*='#'] {
  stroke: inherit;
  fill: none;
}

.c7 *[fill-rule],
.c7 *[FILL-RULE],
.c7 *[fill*='#'],
.c7 *[FILL*='#'] {
  fill: inherit;
  stroke: none;
}

.c15 {
  display: inline-block;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 24px;
  height: 24px;
  fill: #7D4CDB;
  stroke: #7D4CDB;
}

.c15 g {
  fill: inherit;
  stroke: inherit;
}

.c15 *:not([stroke])[fill='none'] {
  stroke-width: 0;
}

.c15 *[stroke*='#'],
.c15 *[STROKE*='#'] {
  stroke: inherit;
  fill: none;
}

.c15 *[fill-rule],
.c15 *[FILL-RULE],
.c15 *[fill*='#'],
.c15 *[FILL*='#'] {
  fill: inherit;
  stroke: none;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c6 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  color: inherit;
  outline: none;
  border: none;
  padding: 0;
  text-align: inherit;
  opacity: 0.3;
  cursor: default;
  height: 100%;
  line-height: 0;
}

cursor:default .c5:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

cursor:default .c5:focus > circle,
cursor:default .c6:focus > ellipse,
cursor:default .c6:focus > line,
cursor:default .c6:focus > path,
cursor:default .c6:focus > polygon,
cursor:default .c6:focus > polyline,
cursor:default .c6:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

cursor:default .c5:focus::-moz-focus-inner {
  border: 0;
}

.c6:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c6:focus:not(:focus-visible) > circle,
.c6:focus:not(:focus-visible) > ellipse,
.c6:focus:not(:focus-visible) > line,
.c6:focus:not(:focus-visible) > path,
.c6:focus:not(:focus-visible) > polygon,
.c6:focus:not(:focus-visible) > polyline,
.c6:focus:not(:focus-visible) > rect {
  outline: none;
  box-shadow: none;
}

.c6:focus:not(:focus-visible)::-moz-focus-inner {
  border: 0;
}

.c14 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  color: inherit;
  outline: none;
  border: none;
  padding: 0;
  text-align: inherit;
  line-height: 0;
  padding: 12px;
}

.c14:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c14:focus > circle,
.c14:focus > ellipse,
.c14:focus > line,
.c14:focus > path,
.c14:focus > polygon,
.c14:focus > polyline,
.c14:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c14:focus::-moz-focus-inner {
  border: 0;
}

.c14:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c14:focus:not(:focus-visible) > circle,
.c14:focus:not(:focus-visible) > ellipse,
.c14:focus:not(:focus-visible) > line,
.c14:focus:not(:focus-visible) > path,
.c14:focus:not(:focus-visible) > polygon,
.c14:focus:not(:focus-visible) > polyline,
.c14:focus:not(:focus-visible) > rect {
  outline: none;
  box-shadow: none;
}

.c14:focus:not(:focus-visible)::-moz-focus-inner {
  border: 0;
}

.c17 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  color: inherit;
  outline: none;
  border: none;
  padding: 0;
  text-align: inherit;
  height: 100%;
  line-height: 0;
}

.c17:hover {
  background-color: rgba(221,221,221,0.4);
  color: #000000;
}

.c17:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c17:focus > circle,
.c17:focus > ellipse,
.c17:focus > line,
.c17:focus > path,
.c17:focus > polygon,
.c17:focus > polyline,
.c17:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c17:focus::-moz-focus-inner {
  border: 0;
}

.c17:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c17:focus:not(:focus-visible) > circle,
.c17:focus:not(:focus-visible) > ellipse,
.c17:focus:not(:focus-visible) > line,
.c17:focus:not(:focus-visible) > path,
.c17:focus:not(:focus-visible) > polygon,
.c17:focus:not(:focus-visible) > polyline,
.c17:focus:not(:focus-visible) > rect {
  outline: none;
  box-shadow: none;
}

.c17:focus:not(:focus-visible)::-moz-focus-inner {
  border: 0;
}

.c2 {
  position: relative;
  overflow: hidden;
}

.c8 {
  visibility: visible;
  position: relative;
  width: 100%;
  height: 100%;
  overflow: hidden;
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
}

.c9 {
  visibility: hidden;
  position: absolute;
  width: 100%;
  height: 100%;
  overflow: hidden;
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
}

.c4 {
  position: absolute;
  z-index: 1;
  left: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c11 {
  position: absolute;
  z-index: 1;
  bottom: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c16 {
  position: absolute;
  z-index: 1;
  right: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

<div
    class="c0"
  >
    <div
      class="c1 c2"
    >
      <div
        class="c3 c4"
      >
        <button
          aria-label="Go to slide 0"
          class="c5 c6"
          disabled=""
          type="button"
        >
          <svg
            aria-label="Previous"
            class="c7"
            viewBox="0 0 24 24"
          >
            <path
              d="M17 2 7 12l10 10"
              fill="none"
              stroke="#000"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
      <div
        class="c1 c8"
      >
        <div
          class="c1"
        >
          <div>
            Slide One
          </div>
        </div>
      </div>
      <div
        class="c1 c9"
      >
        <div
          class="c1"
        >
          <div>
            Slide Two
          </div>
        </div>
      </div>
      <div
        class="c10 c11"
      >
        <div
          class="c12"
        >
          <div
            class="c13"
          >
            <button
              aria-label="Jump to slide 1"
              class="c5 c14"
              type="button"
            >
              <svg
                aria-label="Subtract"
                class="c15"
                viewBox="0 0 24 24"
              >
                <path
                  d="M2 12h20"
                  fill="none"
                  stroke="#000"
                  stroke-width="2"
                />
              </svg>
            </button>
            <button
              aria-label="Jump to slide 2"
              class="c5 c14"
              type="button"
            >
              <svg
                aria-label="Subtract"
                class="c7"
                viewBox="0 0 24 24"
              >
                <path
                  d="M2 12h20"
                  fill="none"
                  stroke="#000"
                  stroke-width="2"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="c3 c16"
      >
        <button
          aria-label="Go to slide 2"
          class="c5 c17"
          type="button"
        >
          <svg
            aria-label="Next"
            class="c7"
            viewBox="0 0 24 24"
          >
            <path
              d="m7 2 10 10L7 22"
              fill="none"
              stroke="#000"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Carousel controlled component 1`] = `
<DocumentFragment>
  .c7 {
  display: inline-block;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 24px;
  height: 24px;
  fill: #666666;
  stroke: #666666;
}

.c7 g {
  fill: inherit;
  stroke: inherit;
}

.c7 *:not([stroke])[fill='none'] {
  stroke-width: 0;
}

.c7 *[stroke*='#'],
.c7 *[STROKE*='#'] {
  stroke: inherit;
  fill: none;
}

.c7 *[fill-rule],
.c7 *[FILL-RULE],
.c7 *[fill*='#'],
.c7 *[FILL*='#'] {
  fill: inherit;
  stroke: none;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
}

.c6 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  color: inherit;
  outline: none;
  border: none;
  padding: 0;
  text-align: inherit;
  height: 100%;
  line-height: 0;
}

.c6:hover {
  background-color: rgba(221,221,221,0.4);
  color: #000000;
}

.c6:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c6:focus > circle,
.c6:focus > ellipse,
.c6:focus > line,
.c6:focus > path,
.c6:focus > polygon,
.c6:focus > polyline,
.c6:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c6:focus::-moz-focus-inner {
  border: 0;
}

.c6:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c6:focus:not(:focus-visible) > circle,
.c6:focus:not(:focus-visible) > ellipse,
.c6:focus:not(:focus-visible) > line,
.c6:focus:not(:focus-visible) > path,
.c6:focus:not(:focus-visible) > polygon,
.c6:focus:not(:focus-visible) > polyline,
.c6:focus:not(:focus-visible) > rect {
  outline: none;
  box-shadow: none;
}

.c6:focus:not(:focus-visible)::-moz-focus-inner {
  border: 0;
}

.c11 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  color: inherit;
  outline: none;
  border: none;
  padding: 0;
  text-align: inherit;
  opacity: 0.3;
  cursor: default;
  height: 100%;
  line-height: 0;
}

cursor:default .c5:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

cursor:default .c5:focus > circle,
cursor:default .c11:focus > ellipse,
cursor:default .c11:focus > line,
cursor:default .c11:focus > path,
cursor:default .c11:focus > polygon,
cursor:default .c11:focus > polyline,
cursor:default .c11:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

cursor:default .c5:focus::-moz-focus-inner {
  border: 0;
}

.c11:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c11:focus:not(:focus-visible) > circle,
.c11:focus:not(:focus-visible) > ellipse,
.c11:focus:not(:focus-visible) > line,
.c11:focus:not(:focus-visible) > path,
.c11:focus:not(:focus-visible) > polygon,
.c11:focus:not(:focus-visible) > polyline,
.c11:focus:not(:focus-visible) > rect {
  outline: none;
  box-shadow: none;
}

.c11:focus:not(:focus-visible)::-moz-focus-inner {
  border: 0;
}

.c2 {
  position: relative;
  overflow: hidden;
}

.c8 {
  visibility: hidden;
  position: relative;
  width: 100%;
  height: 100%;
  overflow: hidden;
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
}

.c9 {
  visibility: visible;
  position: absolute;
  width: 100%;
  height: 100%;
  overflow: hidden;
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
}

.c4 {
  position: absolute;
  z-index: 1;
  left: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c10 {
  position: absolute;
  z-index: 1;
  right: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

<div
    class="c0"
  >
    <div
      class="c1 c2"
    >
      <div
        class="c3 c4"
      >
        <button
          aria-label="Go to slide 1"
          class="c5 c6"
          type="button"
        >
          <svg
            aria-label="Previous"
            class="c7"
            viewBox="0 0 24 24"
          >
            <path
              d="M17 2 7 12l10 10"
              fill="none"
              stroke="#000"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
      <div
        class="c1 c8"
      >
        <div
          class="c1"
        >
          <div>
            Slide One
          </div>
        </div>
      </div>
      <div
        class="c1 c9"
      >
        <div
          class="c1"
        >
          <div>
            Slide Two
          </div>
        </div>
      </div>
      <div
        class="c3 c10"
      >
        <button
          aria-label="Go to slide 3"
          class="c5 c11"
          disabled=""
          type="button"
        >
          <svg
            aria-label="Next"
            class="c7"
            viewBox="0 0 24 24"
          >
            <path
              d="m7 2 10 10L7 22"
              fill="none"
              stroke="#000"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
